signer/src/testing/transaction_signer.rs

```diff
@@ -0,0 +1,268 @@
+            .map(|_| {
+                let event_loop_harness = EventLoopHarness::create(
+                    &mut rng,
+                    network.connect(),
+                    (self.storage_constructor)(),
+                    self.context_window,
+                );
+
+                event_loop_harness.start()
+            })
+            .collect();
+
+        for handle in event_loop_handles.iter_mut() {
+            Self::write_test_data(&test_data, &mut handle.storage).await;
+        }
+
+        for handle in event_loop_handles.iter() {
+            handle
+                .notification_tx
+                .send(())
+                .expect("failed to send notification");
+        }
+
+        // TODO(258): Ensure we can wait for the signers to finish processing messages
+        tokio::time::sleep(std::time::Duration::from_millis(500)).await;
+
+        for handle in event_loop_handles {
+            let storage = handle.stop_event_loop().await;
+            let context_window_block_hashes =
+                self.extract_context_window_block_hashes(&storage).await;
+
+            Self::assert_only_deposit_requests_in_context_window_has_decisions(
+                &storage,
+                &context_window_block_hashes,
+                &test_data.deposit_requests,
+                self.num_signers,
+            )
+            .await;
+        }
+    }
+
+    async fn write_test_data(test_data: &testing::storage::model::TestData, storage: &mut S) {
+        test_data.write_to(storage).await;
+    }
+
+    async fn extract_context_window_block_hashes(
+        &self,
+        storage: &S,
+    ) -> Vec<model::BitcoinBlockHash> {
+        let mut context_window_block_hashes = Vec::new();
+        let mut block_hash = storage
+            .get_bitcoin_canonical_chain_tip()
+            .await
+            .unwrap()
+            .expect("found no canonical chain tip");
+
+        for _ in 0..self.context_window {
+            context_window_block_hashes.push(block_hash.clone());
+            let Some(block) = storage.get_bitcoin_block(&block_hash).await.unwrap() else {
+                break;
+            };
+            block_hash = block.parent_hash;
+        }
+
+        context_window_block_hashes
+    }
+
+    async fn assert_only_deposit_requests_in_context_window_has_decisions(
+        storage: &S,
+        context_window_block_hashes: &[model::BitcoinBlockHash],
+        deposit_requests: &[model::DepositRequest],
+        num_expected_decisions: usize,
+    ) {
+        for deposit_request in deposit_requests {
+            let signer_decisions = storage
+                .get_deposit_signers(&deposit_request.txid, deposit_request.output_index)
+                .await
+                .unwrap();
+
+            for deposit_request_block in storage
+                .get_bitcoin_blocks_with_transaction(&deposit_request.txid)
+                .await
+                .unwrap()
+            {
```

Comment:
nit: I find having things broken out like this a bit easier to read:
```rust
let blocks = storage
    .get_bitcoin_blocks_with_transaction(&deposit_request.txid)
    .await
    .unwrap();

for deposit_request_block in blocks {
...
```

Reply:
Yeah I agree. Updated ✅